### PR TITLE
[gui] GGUI 5/n: swap chain and app context

### PR DESCRIPTION
**Pull request by AmesingFlank**
Related issue = #2646 

This is the 5th of a series of PRs that adds a GPU-based GUI to taichi. This PR adds two classes:
* SwapChain: which manages the vulkan swap chain.
* AppContext: which owns all the vulkan related resources needed by the GUI. This includes a `SwapChain`, an `EmbeddedVulkanDevice`, and for now a render pass.

### CODE REVIEW

File: taichi/ui/backends/vulkan/app_context.cpp [hunk 1]
```diff
@@ -0,0 +1,134 @@
...
+#include "taichi/ui/backends/vulkan/swap_chain.h"
+
+TI_UI_NAMESPACE_BEGIN
+
+namespace vulkan {
+
+using namespace taichi::lang::vulkan;
+
+// EmbeddedVulkanDevice will check that these are supported
+const std::vector<std::string> device_extensions = {
```
Comment (k-ye) [inline taichi/ui/backends/vulkan/app_context.cpp:13]: nit:

1. enclose this with anonymouse namespace
2. `constexpr std::string_view kDeviceExtensions[] = { ... }`

File: taichi/ui/backends/vulkan/app_context.h [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+  VkInstance instance() const;
+  VkDevice device() const;
+  VkPhysicalDevice physical_device() const;
+  taichi::lang::vulkan::VulkanQueueFamilyIndices queue_family_indices() const;
+  VkQueue graphics_queue() const;
+  VkQueue present_queue() const;
+  VkCommandPool command_pool() const;
+  VkRenderPass render_pass() const;
+
+  AppConfig config;
```
Comment (k-ye) [inline taichi/ui/backends/vulkan/app_context.h:32]: Hide these if you don't intend other people to write them?

File: taichi/ui/backends/vulkan/app_context.h [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+  SwapChain swap_chain;
+
+  GLFWwindow *glfw_window;
+
+ private:
+  std::unique_ptr<taichi::lang::vulkan::EmbeddedVulkanDevice> vulkan_device_;
```
Comment (k-ye) [inline taichi/ui/backends/vulkan/app_context.h:38]: init

File: taichi/ui/backends/vulkan/swap_chain.h [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+  VkFormat swap_chain_image_format;
+  VkExtent2D swap_chain_extent;
+  std::vector<VkImageView> swap_chain_image_views;
+  std::vector<VkFramebuffer> swap_chain_framebuffers;
+
+  VkImage depth_image;
+  VkDeviceMemory depth_image_memory;
+  VkImageView depth_image_view;
+
+  std::vector<VkSemaphore> image_available_semaphores;
```
Comment (k-ye) [inline taichi/ui/backends/vulkan/swap_chain.h:23]: Do these member variables need to be public?

**Review by k-ye — COMMENTED** [review]
Great!

File: taichi/ui/backends/vulkan/app_context.h (hunk 2 above)
Comment (AmesingFlank) [reply taichi/ui/backends/vulkan/app_context.h:35]: config is written to by others when window is resized. I'm hiding everything else.

**Review by k-ye — APPROVED** [review]

**Pull request closed without merging**